CAM: fix handling of Active state and CoolantMode with nested dressups

jffmichi opened the pull request:
Previously the refactored postprocessor was only checking the operation itself and it's `Base` for the `Active` state and  `CoolantMode`. This breaks when multiple dressups are appied to an operation.

luzpaz commented:
cc @LarryWoestman JFYI

LarryWoestman reviewed (approved):
This looks good to me.

LarryWoestman commented:
It would be nice to have at least one test for this change.

jffmichi commented:
As far as I can tell there are currently no tests for any postprocessor that use dressups... so I don't really know where to start. I'll have a look...

LarryWoestman commented:
I don't know of any tests for dressups, either :-(

All I can think of is to create a job with an appropriate dressup, run the refactored test postprocessor on the job, and check a few of the most relevant lines of gcode to verify if they show the correct gcode.
There are a fair number of tests for the refactored test postprocessor that might give you some ideas.  I would also be glad to help you figure something out if you wish.

jffmichi commented:
I noticed that the refactored postprocessors handle the pre- and post-operation comments differently. Fixed that and updated the affected tests. Is it OK to include this in this PR?

Also added an unit test to cover the changes in the first commit of this PR. @LarryWoestman could you have a look and let me know if anything needs changing?

LarryWoestman commented:
I agree that the pre- and post- operation comments should work the same.  Ideally this would be a separate PR but it is simple enough to leave it as part of this PR in this case, in my opinion.

It would be nice to also change the comment where SHOW_OPERATION_LABEL is initialized to mention that it works for both pre- and post- operation comments.  That is where the most documentation about what the various values mean is.

I like the tests for the dressups!  Well done!

I have reviewed all of the code changes and approve of it.

jffmichi commented:
Comment for SHOW_OPERATION_LABELS updated. pre-commit.com changes integrated into the respective commits.

sliptonic commented:
[simple.zip](https://github.com/user-attachments/files/20192247/simple.zip)

The attached file has a simple millface with coolant enabled.  The op has two dressups.  With only one dressup, the coolant mode is preserved.  With two, it is lost.

There's a helper function to find the base op.  It will handle multiple levels of dressup.

```
import Path.Dressup.Utils as u

obj = doc.getObject("LeadInOutDressup")
b = u.baseOp(obj)
b.Name
'MillFace'
```

jffmichi commented:
It seems you're using the old postprocessor "linuxcnc"? The changes and tests in this PR only affect the new "refactored_***" postprocessors (and "generic" and "snapmaker").

There are two sets of these helper functions:
a) `baseOp` and `toolController` from Path.Dressup.Utils
b) `opProperty`, `toolControllerForOp` and the newly added `coolantModeForOp` and `activeForOp` from Path.Base.Util

Those two sets of functions do basically the same, although with different implementations. `baseOp` in the a)-set checks if the name of the object contains "Dressup". `opProperty` in the b)-set checks if the object has a "Base" attribute. Since the new base class `PostProcessor` was already using `toolControllerForOp` I chose to stick with the b)-set.

Also, checking for a string in the object name in seems a little fishy to me...

jffmichi commented:
The blocks of code that read the active state and coolant mode seem to be the same (or very similar) in almost all "old" postprocessors. Should I try to fix those too? It should be straightforward but I'm not sure I want to write tests for all those...

sliptonic commented:
@LarryWoestman  How close do you think we are to deprecating the old posts?   I don't want to invest energy in fixing them if the refactored posts are functionally equivalent or better.  My suspicion is that most people are still running the old ones and we should 'encourage' them to move on.

sliptonic commented:
> The blocks of code that read the active state and coolant mode seem to be the same (or very similar) in almost all "old" postprocessors. Should I try to fix those too? It should be straightforward but I'm not sure I want to write tests for all those...

The whole reason for Larry's refactor work is to simplify and get to a testable framework.  So I don't think you should spend the time to write tests.  If you want to fix the old posts, that would be great but would also give people a reason to stick with them rather than moving to the new.

LarryWoestman commented:
I would like to deprecate the old posts, but there is still a lot of work left to finish getting the refactored code base to have "enough" customizability.  Then there are a number of old posts that need to be re-done as refactored posts, which most likely will cause more refactoring to occur.

At my current rate of progress my guess is that it will take at least two or three more years to get "close" to be ready to start deprecating most of the old posts.  There may be some old posts that are so different from the refactored code base that they can't be redone on the refactored code base without major changes.  There may also be some old posts that aren't being used any more or aren't worth the work to redo them.

luzpaz commented:
@sliptonic maybe worth adding what post-processors are running to the About info so we can see the telemetry on this ?

LarryWoestman commented:
The four posts that I started from (grbl, linuxcnc, centroid, and mach3mach4) could be deprecated sooner rather than later if we wish.

LarryWoestman commented:
@luzpaz We don't know what posprocessor (or postprocessors) are being used until the postprocessor is chosen at run time.  Some of the FreeCAD files will have information about the last postprocessor that was used for a given job, but we don't know what other postprocessors might have been used as well.  I suspect that many people only have one cnc system and only use one postprocessor, but that is just a guess on my part.

LarryWoestman commented:
I'm trying to resist the temptation to fix old postprocessors "just because", but rather wait and see if someone is asking for the fix.  It is the only way I can figure out to find out which postprocessors are still being used.  In some cases the existing postprocessor code is working for the existing uses, and putting time into fixing them may not have a high return on the time and effort it takes.

I would be happier if I had working refactored versions of most of the postprocessors to offer, but (except for the four I listed before and a couple more that were written for the first time based on the refactored code base) that is work that still needs to be done.

LarryWoestman commented:
Also, in many cases, the more time and effort I spend on fixing old postprocessors the less time and energy I have available to spend on working on their replacements.

jffmichi commented:
Ok, since the proposed changes were mostly search and replace and since the non-refactored postprocessors are probably going to be around for a while I changed the ones that were already checking the Active state and/or CoolantMode. There seem to be still some left that don't check the Active state at all but since operations will output empty gcode if they are deactivated this probably comes down to some extra comments in the resulting gcode file. CoolantMode should be fixed everywhere.

I exported a small test file and manually had a look at it with all postprocessors I touched. I found a few minor issues in the following:
    
uccnc now also checks the Active state.

dynapath_4060 gave an error (also without my modifications):
```
  File "/usr/lib/freecad/Mod/CAM/Path/Post/scripts/dynapath_4060_post.py", line 433, in parse
    + PostUtils.fmt(x, PRECISION, UNITS)
  File "/usr/lib/freecad/Mod/CAM/Path/Post/Utils.py", line 288, in fmt
    fnum = "%.*f" % (dec, num / 25.4)  # since FreeCAD uses metric units internally
```

marlin wrote the output to a file named "-" in the current application directory, then opened the file dialog and write the output to the correct location.
    
rrf only wrote the output to "-".

LarryWoestman commented:
Thanks for fixing the non-refactored postprocessors.  I suspect that some people will benefit from your work.  I will review the additional changes when I have time, hopefully in the next couple of days.

LarryWoestman reviewed (approved):
I have reviewed these changes and they look reasonable to me.

For future reference, the use of "pyopen" instead of "open" appears to be left over from a long time ago and doesn't appear to be needed any more.  I took it out of the refactored* code base quite a while ago without any known issues.  It isn't worth "fixing" unless you are changing that part of the code for other reasons (in my opinion).
